Connect Item Directions GraphQl to UI

aimee-mcneil-melville opened the pull request:
<!-- IMPORTANT!
  - Every PR must reference an issue; this helps to explain the intent of the PR
 -->

Fixes #5648 

# 👩🏻‍💻 What does this PR do?

<!-- Explain the changes you made -->
This connects the graphQl to the front end, receiving item direction data to the prescription form.

Item directions will come from the table, through to the UI where a user will see a list of abbreviated default directions when clicking in the dropdown. Clicking one will fill the directions text box with the expanded version of the abbreviation.
If no defaults are configured, this will still drop down but display 'No default directions available'

<img width="563" alt="Screenshot 2025-02-10 at 12 44 08 PM" src="https://github.com/user-attachments/assets/cc48c689-525e-43ab-9a37-d983af684ab4" />

<img width="561" alt="Screenshot 2025-02-10 at 12 40 02 PM" src="https://github.com/user-attachments/assets/b185d1e3-4e1e-425b-a3aa-2265c3e42c6d" />

Other changes:
Added itemDirections to the item nodes. As prescriptions used the 'StockOut' fragment, adding the itemDirections to Stock Out required adding through Outbound Shipment and everywhere that was connected to Stock Out

Therefore, a PresciptionLine fragment has been created replicating StockOut, with the addition of item directions. This will allow future distinctions between StockOut and PrescriptionLine in future as well.
Throughout the prescriptions area, types have been updated from DraftStockOutLine to PrescriptionLine
New utils/tests created for prescriptions - duplicated from the stock out utils. I removed prescription specific utils from the stock out file, but kept all stock out utils/tests on the prescription file as many changes will need to be done both, which the tests should catch

<!-- why are the changes needed -->

<!-- Add a screenshot if there are UI changes  -->

## 💌 Any notes for the reviewer?

Can be tested along with #6421 

<!-- Do you have any specific questions for the reviewer? -->

<!-- Is there a high risk/complicated change they should focus on? -->

<!-- any general areas of the codebase touched? any side effects caused? -->

<!-- Anything half cooked but going to be finished off in a different PR? -->

# 🧪 Testing

Go to Prescriptions -> New Prescription -> Search for a Patient
Add Item -> Choose an item and quantity

For an item without default direction options configured:
- [ ] The dropdown will be clickable, and display 'No default directions available' in the dropdown. Clicking on this has no action 

For an item with default direction options configured:
- [ ] In the `Item default directions` dropdown, select a direction abbreviation
- [ ]  See that this enters a direction into the Directions field
- [ ]  Edit the directions field directly - the selected option will clear
- [ ]  Select from the dropdown again, then enter any text into the abbreviations field
- [ ]  Tab or click out of the field, see the typed abbreviation remain and the selected option will clear
- [ ]  Click Save -> go to another screen and navigate back to the same prescription - see that the direction is saved
- [ ]  Automated tests will pass

# 📃 Documentation

- [x] **Part of an epic**: documentation will be completed for the feature as a whole
- [ ] **No documentation required**: no user facing changes or a bug fix which isn't a change in behaviour
- [ ] **These areas should be updated or checked**: <!-- _(e.g.)_ New `issued` column in `Requisitions` indicates stock quantity already in shipments -->
  1.
  2.

Chris-Petty commented:
Forgive the pr base chop, was fiddling cause I swear there was overlap but I was wrong I think, or I'm doing it wrong. `git diff` comparisons there is definitely overlap...

Chris-Petty reviewed (changes requested):
Overall the code looks good, epic effort! Unfortunately the feature seems to have regressed though, when I select a default it does not populate the directions and it gets in some tricky state where entering abbreviations directly stops working, and the directions input doesn't update with any of my keystrokes either:


https://github.com/user-attachments/assets/74376655-ad89-4ff3-a23c-c51ee4464d5e

aimee-mcneil-melville commented:
This seems to be related to #6214 and #6248 - I have merged the fix however still occuring. Whenever the stock unallocates, the directions and abbreviations state is still updating but not displayed in the text area. Selecting a default direction will most of the time cause the stock to unallocate

https://github.com/user-attachments/assets/578b0122-e2ea-4dbc-836c-6bde7b6796c6

Chris-Petty commented:
@aimee-mcneil-melville State is still getting nuked - should I be re-reviewing this yet? 😁

aimee-mcneil-melville commented:
Working on it with @lache-melvin right now!

Chris-Petty commented on client/packages/invoices/src/Prescriptions/LineEditView/PrescriptionLineEditForm.tsx:
```diff
@@ -237,25 +237,25 @@ export const PrescriptionLineEditForm: React.FC<
+      setDefaultDirection(direction);
+      const note = getPrescriptionDirections(direction, options);
+      updateNotes(note);
+      setAbbreviation('');
```
When you set the directions, it could put into the abbreviation input similar to OG, rather than clearing the abbreviation input. Not a biggy though in the spirit of getting this merged!

Chris-Petty commented on client/packages/invoices/src/Prescriptions/LineEditView/PrescriptionLineEditForm.tsx:
```diff
@@ -237,25 +237,25 @@ export const PrescriptionLineEditForm: React.FC<
+      updateNotes(note);
+      setAbbreviation('');
+    },
+    [defaultDirection]
```
I'm not sure why we need this dependency but I haven't used useCallback much myself...

Chris-Petty replied:
Actually I'm getting suspicious of it... yea I tested just removing `useCallback` and now the state isn't cleared and all is well... yup so you actually, shockingly, have a dependency on `updateNotes` the function as it is memoized using `useCallback` as well so in this function you were calling an old version that hadn't got re-evaluated with the draft lines in it. Yikes, it's a minefield out there...

I'd honestly be happy with just not wrapping this one in useCallback, surely dropdown menu isn't too expensive to rerender? Then again, this situation exists maybe I shouldn't be too optimistic hahahaha.

Chris-Petty reviewed (changes requested):
caching is the devil

aimee-mcneil-melville commented:
Fix is in! Changed the useCallback and added a useEffect to clear state when switching items

Some things to improve on later:
Adding abbreviations, default directions to the DraftPrescriptionLine
Not relying on the useEffect for managing the updates when changing items -> either autosave or prompt to save when changing items

Chris-Petty reviewed (approved):
I deem this looking goodeth
